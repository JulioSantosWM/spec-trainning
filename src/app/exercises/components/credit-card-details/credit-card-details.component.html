<mat-card>
  <mat-card-header class="mat-card-header">
    <mat-card-title>Credit Card Details</mat-card-title>
    <mat-card-subtitle>Disclaimer: this is sensible information</mat-card-subtitle>
    <button mat-icon-button
      (click)="toggleExpandCollapse()">
      <mat-icon>{{ expanded ? "unfold_less" : "unfold_more" }}</mat-icon>
    </button>
  </mat-card-header>
  <mat-card-content class="mat-card-content" *ngIf="expanded">
    <ng-container *ngIf="!isLoading; else loading">
      <div>
        <label>Provider:</label>
        <span> {{ creditCardDetails?.credit_card_type }}</span>
      </div>
      <div>
        <label>Number:</label>
        <span> {{ creditCardDetails?.credit_card_number }}</span>
      </div>
      <div>
        <label>Expiry date:</label>
        <span> {{ creditCardDetails?.credit_card_expiry_date }}</span>
      </div>
    </ng-container>
  </mat-card-content>
</mat-card>


<ng-template #loading>
  <div class="spinner-container">
    <mat-spinner></mat-spinner>
  </div>
</ng-template>