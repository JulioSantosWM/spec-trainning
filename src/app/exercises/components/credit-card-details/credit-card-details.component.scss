label {
  padding-right: 12px;
  font-weight: 600;
}
