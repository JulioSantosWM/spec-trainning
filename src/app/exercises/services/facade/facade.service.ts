import { Injectable } from "@angular/core";
import { UsersOpenApiService, CreditCardOpenApiService, TemporaryStoreService } from "../../mocks/facade.mock";
import { firstValueFrom, map } from "rxjs";
import { UserDto } from "../../models/user.dto";
import { CreditCardDto } from "../../models/credit-card.dto";

@Injectable()
export class FacadeService {

  constructor(
    private readonly userOpenApiService: UsersOpenApiService,
    private readonly creditCardOpenApiService: CreditCardOpenApiService,
    private readonly temporaryStoreService: TemporaryStoreService,
  ) {
  }

  public getUsers(): Promise<UserDto[]> {
    return firstValueFrom(this.userOpenApiService.getUsers());
  }

  public getUserDetails(uid: string): Promise<UserDto> {
    return firstValueFrom(this.userOpenApiService.getUser(uid));
  }

  public getPrimaryCreditCard(uid: string): Promise<CreditCardDto> {
    return firstValueFrom(this.creditCardOpenApiService.getPrimaryCreditCard(uid));
  }

  public getAvailableCreditCards(uid: string): Promise<CreditCardDto[]> {
    return firstValueFrom(this.creditCardOpenApiService.getAvailableCreditCards(uid));
  }

  public getIsCCExpanded(): Promise<boolean> {
    return firstValueFrom(
      this.temporaryStoreService.getStoredInformation<boolean>('cc_details_expanded')
        .pipe(map(val => !!val))
    );
  }

  public storeIsCCExpanded(isExpanded: boolean): Promise<void> {
    return this.temporaryStoreService.storeInformation('cc_details_expanded', isExpanded);
  }

}