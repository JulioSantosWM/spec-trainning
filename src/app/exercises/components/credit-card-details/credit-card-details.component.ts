import { CommonModule } from "@angular/common";
import { Component, EventEmitter, Input, OnInit, Output } from "@angular/core";
import { CreditCardDto } from "../../models/credit-card.dto";
import { FacadeService } from "../../services/facade/facade.service";
import { MatProgressSpinnerModule } from "@angular/material/progress-spinner";
import { GlobalStateService } from "../../mocks/global-state.service";
import { filter, map } from "rxjs";
import { MatCardModule } from "@angular/material/card";
import { MatButtonModule } from "@angular/material/button";
import { MatIconModule } from "@angular/material/icon";

@Component({
  selector: 'app-w-credit-card-details',
  templateUrl: './credit-card-details.component.html',
  styleUrls: ['./credit-card-details.component.scss'],
  standalone: true,
  imports: [
    CommonModule,
    MatCardModule,
    MatProgressSpinnerModule,
    MatButtonModule,
    MatIconModule,
  ]
})
export class CreditCardDetailsComponent implements OnInit{
  @Input() expanded!: boolean;
  @Output() onExpandCollapseChange = new EventEmitter<boolean>();

  public creditCardDetails: CreditCardDto | null = null;
  public isLoading = true;

  constructor(
    private facadeService: FacadeService,
    private globalStateService: GlobalStateService,
  ) { }

  public async ngOnInit(): Promise<void> {
    this.globalStateService.selectedUser$
      .pipe(map(selectedUser => selectedUser?.uid || null), filter(val => typeof val === 'string'))
      .subscribe(async uid => {
        this.isLoading = true;
        this.creditCardDetails = await this.facadeService.getPrimaryCreditCard(uid as string);
        this.isLoading = false;
    })
  }

  public toggleExpandCollapse(): void {
    this.expanded = !this.expanded;
    this.onExpandCollapseChange.emit(this.expanded);
  }
}