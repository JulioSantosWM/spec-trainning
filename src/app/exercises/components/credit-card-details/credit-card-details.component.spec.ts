import { MockBuilder, MockInstance, MockRender, MockedComponentFixture, ngMocks } from "ng-mocks"
import { CreditCardDetailsComponent } from "./credit-card-details.component"
import { FacadeService } from "../../services/facade/facade.service";
import { GlobalStateService } from "../../mocks/global-state.service";
import { createMock } from "@golevelup/ts-jest";
import { User } from "../../models/user.model";
import { of } from "rxjs";
import { CreditCardDto } from "../../models/credit-card.dto";
import { fakeAsync, tick } from "@angular/core/testing";

interface Config { expanded?: boolean, onExpandCollapseChange?: jest.Mock };

describe('CreditCardDetailsComponent', () => {
  let fixture: MockedComponentFixture<CreditCardDetailsComponent, Config>;
  let component: CreditCardDetailsComponent;
  
  const mockUser = { uid: 'foo' } as User;
  const mockCreditCardDetails = {
    credit_card_type: 'bar', credit_card_number: 'baz', credit_card_expiry_date: 'qux'
  } as CreditCardDto;

  MockInstance.scope();

  describe('when initializing the component', () => {
    MockInstance.scope();

    it('should not view the spinner if the card is collapsed', () => {

    })
    it('should view the spinner if the card is expanded', () => {

    })
  })

  describe('when the component has loaded', () => {
    MockInstance.scope();

    it('should not view the credit card details if the card is collapsed', fakeAsync(async () => {

    }))

    it('should view the credit card details if the card is expanded', fakeAsync(async () => {

    }))

    it('should the information to its parent when the expand/collapse button is clicked', fakeAsync(async () => {

    }))
  })
})