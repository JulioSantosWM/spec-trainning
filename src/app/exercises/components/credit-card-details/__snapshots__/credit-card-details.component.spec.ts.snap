// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreditCardDetailsComponent when initializing the component should not view the spinner if the card is collapsed 1`] = `
<div
  id="root0"
>
  <app-w-credit-card-details>
    <mat-card>
      <mat-card-header
        class="mat-card-header"
      >
        <mat-card-title>
          Credit Card Details
        </mat-card-title>
        
        <mat-card-subtitle>
          Disclaimer: this is sensible information
        </mat-card-subtitle>
        
        <button
          mat-icon-button=""
        >
          <mat-icon>
            unfold_more
          </mat-icon>
        </button>
      </mat-card-header>
      
    </mat-card>
    
  </app-w-credit-card-details>
</div>
`;

exports[`CreditCardDetailsComponent when initializing the component should view the spinner if the card is expanded 1`] = `
<div
  id="root1"
>
  <app-w-credit-card-details>
    <mat-card>
      <mat-card-header
        class="mat-card-header"
      >
        <mat-card-title>
          Credit Card Details
        </mat-card-title>
        
        <mat-card-subtitle>
          Disclaimer: this is sensible information
        </mat-card-subtitle>
        
        <button
          mat-icon-button=""
        >
          <mat-icon>
            unfold_less
          </mat-icon>
        </button>
      </mat-card-header>
      <mat-card-content
        class="mat-card-content"
      >
        <div
          class="spinner-container"
        >
          <mat-spinner />
        </div>
        
      </mat-card-content>
      
      
    </mat-card>
    
  </app-w-credit-card-details>
</div>
`;

exports[`CreditCardDetailsComponent when the component has loaded should not view the credit card details if the card is collapsed 1`] = `
<div
  id="root2"
>
  <app-w-credit-card-details>
    <mat-card>
      <mat-card-header
        class="mat-card-header"
      >
        <mat-card-title>
          Credit Card Details
        </mat-card-title>
        
        <mat-card-subtitle>
          Disclaimer: this is sensible information
        </mat-card-subtitle>
        
        <button
          mat-icon-button=""
        >
          <mat-icon>
            unfold_more
          </mat-icon>
        </button>
      </mat-card-header>
      
    </mat-card>
    
  </app-w-credit-card-details>
</div>
`;

exports[`CreditCardDetailsComponent when the component has loaded should view the credit card details if the card is expanded 1`] = `
<div
  id="root3"
>
  <app-w-credit-card-details>
    <mat-card>
      <mat-card-header
        class="mat-card-header"
      >
        <mat-card-title>
          Credit Card Details
        </mat-card-title>
        
        <mat-card-subtitle>
          Disclaimer: this is sensible information
        </mat-card-subtitle>
        
        <button
          mat-icon-button=""
        >
          <mat-icon>
            unfold_less
          </mat-icon>
        </button>
      </mat-card-header>
      <mat-card-content
        class="mat-card-content"
      >
        <div>
          <label>
            Provider:
          </label>
          <span>
             bar
          </span>
        </div>
        <div>
          <label>
            Number:
          </label>
          <span>
             baz
          </span>
        </div>
        <div>
          <label>
            Expiry date:
          </label>
          <span>
             qux
          </span>
        </div>
        
        
      </mat-card-content>
      
      
    </mat-card>
    
  </app-w-credit-card-details>
</div>
`;
